src/combinatorics/additive/salem_spencer.lean

```diff
@@ -0,0 +1,410 @@
+(mul_roth_number_le _).antisymm $ hs.le_mul_roth_number $ subset.refl _
+
+@[to_additive]
+lemma mul_roth_number_empty : mul_roth_number (∅ : finset α) = 0 :=
+nat.eq_zero_of_le_zero $ (mul_roth_number_le _).trans card_empty.le
+
+@[to_additive]
+lemma mul_roth_number_singleton (a : α) : mul_roth_number ({a} : finset α) = 1 :=
+begin
+  convert mul_salem_spencer.roth_number_eq _,
+  rw coe_singleton,
+  exact mul_salem_spencer_singleton a,
+end
+
+@[to_additive]
+lemma mul_roth_number_union_le (s t : finset α) :
+  mul_roth_number (s ∪ t) ≤ mul_roth_number s + mul_roth_number t :=
+let ⟨u, hus, hcard, hu⟩ := mul_roth_number_spec (s ∪ t) in
+calc
+  mul_roth_number (s ∪ t)
+      = u.card : hcard.symm
+  ... = (u ∩ s ∪ u ∩ t).card
+      : by rw [←inter_distrib_left, (inter_eq_left_iff_subset _ _).2 hus]
+  ... ≤ (u ∩ s).card + (u ∩ t).card : card_union_le _ _
+  ... ≤ mul_roth_number s + mul_roth_number t
+      : add_le_add ((hu.mono $ inter_subset_left _ _).le_mul_roth_number $ inter_subset_right _ _)
+          ((hu.mono $ inter_subset_left _ _).le_mul_roth_number $ inter_subset_right _ _)
+
+@[to_additive]
+lemma le_mul_roth_number_product (s : finset α) (t : finset β) :
+  mul_roth_number s * mul_roth_number t ≤ mul_roth_number (s.product t) :=
+begin
+  obtain ⟨u, hus, hucard, hu⟩ := mul_roth_number_spec s,
+  obtain ⟨v, hvt, hvcard, hv⟩ := mul_roth_number_spec t,
+  rw [←hucard, ←hvcard, ←card_product],
+  refine mul_salem_spencer.le_mul_roth_number _ (product_subset_product hus hvt),
+  rw coe_product,
+  exact hu.prod hv,
+end
+
+@[to_additive]
+lemma mul_roth_number_lt_of_forall_not_mul_salem_spencer
+  (h : ∀ t ∈ powerset_len n s, ¬mul_salem_spencer ((t : finset α) : set α)) :
+  mul_roth_number s < n :=
+begin
+  obtain ⟨t, hts, hcard, ht⟩ := mul_roth_number_spec s,
+  rw [←hcard, ←not_le],
+  intro hn,
+  obtain ⟨u, hut, rfl⟩ := exists_smaller_set t n hn,
+  exact h _ (mem_powerset_len.2 ⟨hut.trans hts, rfl⟩) (ht.mono hut),
+end
+
+end monoid
+
+section cancel_comm_monoid
+variables [cancel_comm_monoid α] (s : finset α) (a : α)
+
+@[simp, to_additive] lemma mul_roth_number_map_mul_left :
+  mul_roth_number (s.map $ mul_left_embedding a) = mul_roth_number s :=
+begin
+  refine le_antisymm _ _,
+  { obtain ⟨u, hus, hcard, hu⟩ := mul_roth_number_spec (s.map $ mul_left_embedding a),
+    rw subset_map_iff at hus,
+    obtain ⟨u, hus, rfl⟩ := hus,
+    rw coe_map at hu,
+    rw [←hcard, card_map],
+    exact (mul_salem_spencer_mul_left_iff.1 hu).le_mul_roth_number hus },
+  { obtain ⟨u, hus, hcard, hu⟩ := mul_roth_number_spec s,
+    have h : mul_salem_spencer (u.map $ mul_left_embedding a : set α),
+    { rw coe_map,
+      exact hu.mul_left },
+    convert h.le_mul_roth_number (map_subset_map.2 hus),
+    rw [card_map, hcard] }
+end
+
+@[simp, to_additive] lemma mul_roth_number_map_mul_right :
+  mul_roth_number (s.map $ mul_right_embedding a) = mul_roth_number s :=
+begin
+  convert mul_roth_number_map_mul_left s a,
+  ext,
```

Comment:
Seems like `mul_left_embedding a = mul_right_embedding a` should be a lemma for `cancel_comm_monoids`